docs/puzzles/day2.md

```diff
@@ -0,0 +1,201 @@
+  // we iterate on each entry and move it following the received command
+  val lastPosition = entries.foldLeft(firstPosition)((position, command) => position.move(command))
+  lastPosition.result
+
+case class Position(horizontal: Int, depth: Int):
+  def move(p: Command): Position =
+    p match
+      case Command.Forward(x) => Position(horizontal + x, depth)
+      case Command.Down(x)    => Position(horizontal, depth + x)
+      case Command.Up(x)      => Position(horizontal, depth - x)
+
+  def result = horizontal * depth
+
+enum Command:
+  case Forward(x: Int)
+  case Down(x: Int)
+  case Up(x: Int)
+
+object Command:
+  def from(s: String): Command =
+    s match
+      case s"forward $x" if x.toIntOption.isDefined => Forward(x.toInt)
+      case s"up $x"      if x.toIntOption.isDefined => Up(x.toInt)
+      case s"down $x"    if x.toIntOption.isDefined => Down(x.toInt)
+      case _ => throw new Exception(s"value $s is not valid command")
+```
+
+## Solution of Part 2
+
+The part 2 introduces new rules to move the sonar. 
+So we need a new position that takes into account the `aim` and a new method move with the new rules.
+The remaining code remains the same.
+### Moving the sonar part 2
+```scala
+case class PositionWithAim(horizontal: Int, depth: Int, aim: Int):
+  def move(p: Command): PositionWithAim =
+    p match
+      case Command.Forward(x) => PositionWithAim(horizontal + x, depth + x * aim, aim)
+      case Command.Down(x)    => PositionWithAim(horizontal, depth, aim + x)
+      case Command.Up(x)      => PositionWithAim(horizontal, depth, aim - x)
+
+```
+### Final code for part 2
+```scala
+@main def part2(): Unit =
+  val input = util.Using.resource(Source.fromFile("input/day2"))(_.mkString)
+  val entries: Iterator[Command] = input.linesIterator.map(Command.from)
+  val answer = part2(input)
+
+case class PositionWithAim(horizontal: Int, depth: Int, aim: Int):
+  def move(p: Command): PositionWithAim =
+    p match
+      case Command.Forward(x) => PositionWithAim(horizontal + x, depth + x * aim, aim)
+      case Command.Down(x)    => PositionWithAim(horizontal, depth, aim + x)
+      case Command.Up(x)      => PositionWithAim(horizontal, depth, aim - x)
+
+  def result = horizontal * depth
+
+enum Command:
+  case Forward(x: Int)
+  case Down(x: Int)
+  case Up(x: Int)
+
+object Command:
+  def from(s: String): Command =
+    s match
+      case s"forward $x" if x.toIntOption.isDefined => Forward(x.toInt)
+      case s"up $x"      if x.toIntOption.isDefined => Up(x.toInt)
+      case s"down $x"    if x.toIntOption.isDefined => Down(x.toInt)
+      case _ => throw new Exception(s"value $s is not valid command")
+
+```
+
+## Enum in Scala 3
+An enumeration is used to define a type consisting of a set of named values.
+
+We used to used `sealed trait` with `case class` and `case object` to model this set of named values.
+The syntax is more consise and easier to explain than the combination we had in Scala 2. 
```

Comment:
Scala 3 enums are more concise and easier to read that the Scala 2 ADTs.

```
enum Command:
  case Forward(x: Int)
  case Down(x: Int)
  case Up(x: Int)
```

is almost the same as:
```
sealed trait Command
object Command {
  case class Forward(x: Int) extends Command
  case class Down(x: Int) extends Command
  case class Up(x: Int) extends Command
}
```